### ISSUE TITLE
"Could not fetch ID token (...) socket hang up" in Google Cloud Shell

### ISSUE THREAD

**Issue by ad-m** [googleapis/google-auth-library-nodejs#966]
When I attemp to fetch IDToken within Google Cloud Shell I receive error:

```
> (new (require('google-auth-library').GoogleAuth)).getClient().then(client => client.fetchIdToken("http://google.com/")).then(token => console.log(token))
Promise {
  <pending>,
  domain:
   Domain {
     domain: null,
     _events:
      [Object: null prototype] {
        removeListener: [Function: updateExceptionCapture],
        newListener: [Function: updateExceptionCapture],
        error: [Function: debugDomainError] },
     _eventsCount: 3,
     _maxListeners: undefined,
     members: [] } }
> (node:693) UnhandledPromiseRejectionWarning: FetchError: Could not fetch ID token: request to http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/identity?format=full&audience=http%3A%2F%2Fgoogle.com%2F failed, reason: socket hang up
    at ClientRequest.<anonymous> (/home/zen_vantalye/node_modules/node-fetch/lib/index.js:1455:11)
    at ClientRequest.emit (events.js:182:13)
    at ClientRequest.EventEmitter.emit (domain.js:441:20)
    at Socket.socketOnEnd (_http_client.js:426:9)
    at Socket.emit (events.js:187:15)
    at Socket.EventEmitter.emit (domain.js:459:23)
    at endReadableNT (_stream_readable.js:1094:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
(node:693) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:693) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

When I attemp to fetch IdToken locally using ```GOOGLE_APPLICATION_CREDENTIALS``` I don't have any issue:
```
$ GOOGLE_APPLICATION_CREDENTIALS=billing-enabled-213501-6d7f70c15c52.json nodejs
Welcome to Node.js v12.16.3.

> (new (require('google-auth-library').GoogleAuth)).getClient().then(client => client.fetchIdToken("http://google.com")).then(token => console.log(token))
Promise { <pending> }
>[redacted]
```

From Google Cloud Shell via CLI I am able to obtain IdToken:

```
$ gcloud auth print-identity-token[recacted].(...)
 ```
What I should do? How to obtain Id Token in Google Cloud Shell-compatible way? Is this a library-side error (an exception indicates this) or is the library being used incorrectly?

#### Environment details

  - OS: Google Cloud Shell (production)
  - Node.js version: v10.14.2
  - npm version: 6.4.1
  - `google-auth-library` version: google-auth-library@6.0.0

#### Steps to reproduce

1. Go https://console.cloud.google.com/home/dashboard
2. Open Cloud Shell
3. Execute JavaScript to fetch IdToken.

**Comment by bcoe** [comment]
@ad-m what are you trying to do with the client, I don't think you should need to be running `client.fetchIdToken` manually.

Is there a specific Google API you're attempting to use?

**Comment by ad-m** [comment]
@bcoe , I would like to download the ID token to use it as part of the authentication process – in accordance with OpenID – in my secret distribution system. I want to avoid the use of access token, because I can only verify it in the introspection endpoint token and is vulnerable to theft due to lack of audience restrictions. I want to have the widest possible compatibility of my system (and development tools supporting it). During testing, it turned out that the ID token is available in a small number of Google environments.

**Comment by bcoe** [comment]
@ad-m it seems like your approach could potentially work, but you might do better to hit the [metadata server directly](https://cloud.google.com/compute/docs/storing-retrieving-metadata), or to use the [gcp-metadta](https://github.com/googleapis/gcp-metadata) library.

Have you tried your approach in a GCP environment, such as Cloud Run or Google Cloud Functions, vs., Google Cloud Shell? I wonder if it doesn't have access to the metadata endpoint.

**Comment by ad-m** [comment]
@bcoe, I have tested ```fetchIdToken``` using file that contains the private key (passed via ```GOOGLE_APPLICATION_CREDENTIALS```). It works, because that library have support for self-signed JWT too: https://github.com/googleapis/google-auth-library-nodejs/blob/master/src/auth/jwtclient.ts#L158-L175 In that case - I'd rather use this library correctly - to follow implementation of Application Default Credentials.

**Comment by bcoe** [comment]
@ad-m I just tested this in Cloud Shell, and it seems to be working now:

1. I opened Cloud Shell.
2. I installed the latest version of `googleapis`.
3. I ran `node`, to open a shell.
4. I ran your code `(new (require('google-auth-library').GoogleAuth)).getClient().then(client => client.fetchIdToken("http://google.com/")).then(token => console.log(token))`.

When I ran this code initially, I was given a pop up indicating that Cloud Shell wanted access to credentials; I agreed to this pop up (_I wonder if this might be a new feature_).

Let me know if you continue to bump into issues.
